java-checks/src/main/java/org/sonar/java/checks/BlockingOperationsInVirtualThreadsCheck.java

```diff
@@ -0,0 +1,131 @@
+
+  /*
+   * #### Disclaimers ####
+   *
+   * This rule would benefit from a call graph analysis to determine which blocking operations are executed in a thread. For proper
+   * reporting, it would be important to report issues with a flow, clearly showing the path from the blocking operation to the thread
+   * creation. While we already have tools to find reachable methods (see TreeHelper#findReachableMethodsInSameFile), they do not provide
+   * the necessary information to build a flow and hence a report would be quite confusing. For now, the rule is limited to only very basic
+   * cases.
+   *
+   * On the other hand, this rule may falsely trigger on some cases where the blocking operation is not executed in the thread, but e.g.
+   * returned within a lambda or object to be executed later. I.e. a thread creating a runnable for later execution.
+   */
+
+  private static final String JAVA_LANG_THREAD = "java.lang.Thread";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.NEW_CLASS, Tree.Kind.METHOD_INVOCATION, Tree.Kind.METHOD);
+  }
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava21Compatible();
+  }
+
+  // This rule currently only supports a limited set of blocking operations, focused on core Java HTTP requests.
+  private static final MethodMatchers blockingOperations = MethodMatchers.or(
+    MethodMatchers.create()
+      .ofTypes("java.net.URLConnection", "java.net.HttpURLConnection")
+      .names("getResponseCode", "getResponseMessage")
+      .withAnyParameters()
+      .build(),
+    MethodMatchers.create()
+      .ofTypes(JAVA_LANG_THREAD)
+      .names("sleep")
+      .addParametersMatcher("long")
+      .build());
+
+  private static final MethodMatchers threadCreationConstructors = MethodMatchers.create()
+    .ofSubTypes(JAVA_LANG_THREAD)
+    .constructor()
+    .addParametersMatcher("java.lang.Runnable")
+    .build();
+
+  // For cases where a method is overwritten to define a thread's behavior (instead of e.g. passing a lambda to a Thread's constructor).
+  private static final MethodMatchers overwritableThreadMethods = MethodMatchers.create()
+    .ofSubTypes(JAVA_LANG_THREAD)
+    .names("run")
+    .addWithoutParametersMatcher()
+    .build();
+
+  @Override
+  public void visitNode(Tree tree) {
+    switch (tree.kind()) {
+      case NEW_CLASS -> onConstructorFound((NewClassTree) tree);
+      case METHOD -> onMethodFound((MethodTree) tree);
+    }
+  }
+
+  private void onConstructorFound(NewClassTree newClassTree) {
+    if (threadCreationConstructors.matches(newClassTree)) {
+      analyzeForIssues(newClassTree.arguments().get(0), newClassTree.identifier());
+    }
+  }
+
+  private void onMethodFound(MethodTree tree) {
+    if (overwritableThreadMethods.matches(tree)) {
+      analyzeForIssues(tree, tree.simpleName());
+    }
+  }
+
```

Comment:
And with this fairly simple implementation we should be able to cover the extra constructor!
```

  private void onMethodInvocationFound(MethodInvocationTree tree) {
    var builderType = ((MemberSelectExpressionTree) tree.methodSelect()).expression().symbolType();
    if (threadCreationConstructors.matches(tree) && builderType.is("java.lang.Thread$Builder$OfPlatform")) {
      analyzeForIssues(tree.arguments().get(0), tree.methodSelect());
    }
  }
```